mobile/native/src/ln_dlc/mod.rs

```diff
@@ -560,6 +504,128 @@ fn keep_wallet_balance_and_history_up_to_date(node: &Node) -> Result<()> {
     Ok(())
 }
 
+fn derive_trades_from_filled_orders() -> Result<Vec<WalletHistoryItem>> {
+    let mut trades = vec![];
+    let orders =
+        crate::db::get_filled_orders().context("Failed to get filled orders; skipping update")?;
+
+    match orders.as_slice() {
+        [first, tail @ ..] => {
+            // The first filled order must be an outbound "payment", since coins need to leave the
+            // Lightning wallet to open the first DLC channel.
+            let flow = PaymentFlow::Outbound;
+            let amount_sats = first
+                .trader_margin()
+                .expect("Filled order to have a margin");
+
+            trades.push(WalletHistoryItem {
+                flow,
+                amount_sats,
+                timestamp: first.creation_timestamp.unix_timestamp() as u64,
+                status: Status::Confirmed, // TODO: Support other order/trade statuses
+                wallet_type: WalletHistoryItemType::Trade {
+                    order_id: first.id.to_string(),
+                },
+            });
+
+            let mut total_contracts = match first.direction {
+                Direction::Long => first.quantity,
+                Direction::Short => -first.quantity,
+            };
+            let mut previous_order = first;
+            for order in tail {
+                use trade::Direction::*;
+                let new_contracts = match order.direction {
+                    Long => order.quantity,
+                    Short => -order.quantity,
+                };
+                let updated_total_contracts = total_contracts + new_contracts;
+
+                // Closing the position.
+                if updated_total_contracts == 0.0 {
+                    let open_order = previous_order;
+                    let trader_margin = open_order
+                        .trader_margin()
+                        .expect("Filled order to have a margin");
+                    let execution_price = Decimal::try_from(
+                        order
+                            .execution_price()
+                            .expect("execution price to be set on a filled order"),
+                    )?;
+
+                    let opening_price = open_order
+                        .execution_price()
+                        .expect("initial execution price to be set on a filled order");
+
+                    let pnl = calculations::calculate_pnl(
```

Comment:
Why do we need to re-calculate this every time? Shouldn't this be final if the order was filled?

Reply:
The thing is the PNL is not persisted anywhere in the app (other than when closing a subchannel, but that's also limiting once we allow position resizing via subchannel renewal).

It's definitely not very efficient to have to process this stuff so often, although we were doing the same thing before this PR. We might want to record trades with the relevant metadata as new orders are filled (and executed), so that we can just iterate through those rows here.